### ISSUE TITLE
Calendar Heatmap - Shows unexpected null dates 

### ISSUE THREAD

**Issue by arthurmaenhout** [apache/superset#21870]
I'm making use of the calendar heatmap chart, but i'm getting unexpected null dates for a month outside of the filter. 

#### How to reproduce the bug
1. create heatmap calendar 
2. set filter to 2021-05-01 minus two months (see underneath) 
![image](https://user-images.githubusercontent.com/116160134/196654909-eed1ddb9-9dad-4229-93df-f48c30208234.png)

### Expected results
![image](https://user-images.githubusercontent.com/116160134/196655452-c09a7122-6591-415c-890d-81d2ba7e32a6.png)

### Actual results
But as can be seen on the image underneath, we have a full month with null values
![image](https://user-images.githubusercontent.com/116160134/196654641-320ff8fd-8d4c-4bb4-94e8-5247c2373af3.png)

### Checklist
- [x] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [x] I have reproduced the issue with at least the latest released version of superset.
- [x] I have checked the issue tracker for the same issue and I haven't found one similar.

**Comment by Shazad-khan** [comment]
![image](https://user-images.githubusercontent.com/100838455/226251689-b9393b8c-2d6b-4997-980c-27ce856ac316.png)

It is working as expected please recheck the issue and close this bug

**Comment by rusackas** [comment]
Can anyone confirm if this is still an issue in 3.x?

**Comment by sra** [comment]
This is happening for us in 4.0.2

**Comment by rusackas** [comment]
Yep, this is still real... probably just not the biggest bug and/or not our most highly used chart. I'd encourage anyone to take a look and open a PR if they can solve it!

**Comment by CoolDude937** [comment]
@rusackas hello, I see this is still an issue and I have managed to replicate the issue, although it appears to be slightly different. The month of null values appears to be February 2021 on my end, as opposed to May 2021 in this issue.

<img width="781" height="531" alt="Image" src="https://github.com/user-attachments/assets/9bf17473-7b9c-4359-b2ff-0ee9f4df0ce9" />

I attempted replication on the latest version using the quickstart for Superset, and also the Docker image for version 4.0.2, mentioned to also have this issue. Both times, the “null month” was February 2021. 

Regardless, the general problem seems to persist. Me and a group of others would like to attempt this issue over the next couple of weeks. I hope that is fine with you. Thanks.

**Comment by Bunbunw** [comment]
Hi everyone, I am together with CoolDude937. We tried to figure out why an extra “null” month that is outside of the selected time range is displayed. We suspect that this issue is occurring due to fixed timezone offset variable stdTimezoneOffset in superset-frontend/plugins/legacy-plugin-chart-calendar/src/vendor/cal-heatmap.js, which shifts all timestamps by a fixed value without the consideration of the user's browser timezone offset and DST (Daylight Saving Time) together correctly.
E.g. if your timezone is EST, then its UTC-05:00, because of summer daylight saving time in north america, we have UTC-04:00, which matches our observation where if it's after 4am of the first day of the month, the rendering is correct.
Our intended solution would be to fix the current timezone offset by making sure that the correct value is added to each timestamp with consideration of timezone differences and DST, that way it's minimal and doesn't touch other components. We would also fix more problems related to calendar heatmap if we are done with this issue.
We would like to ask for any feedback or advice regarding our intended solution if possible, thanks!

**Comment by rusackas** [comment]
PR #34564 has been updated with a fix that addresses the root cause of this issue. In addition to the original day-offset bug, the updated PR also resolves the null dates caused by DST phantom hours (the timezone offset issue identified by @CoolDude937 and @Bunbunw in `cal-heatmap.js`). Please take a look and test the fix if you are able to reproduce the issue.

**Comment by rusackas** [comment]
Assuming all is well, and closing this. Holler if there's more to deal with here.
